File: src/matematica/numeros_primos/crivo_eratostenes/crivo_eratostenes.cpp
// ------------------------------------------------------------------
// CRIVO DE ERATÓSTENES
// ------------------------------------------------------------------

#include <iostream>
#include <vector>
using namespace std;

// Função para gerar o array de primos de 0 até n
vector<bool> crivo(int n) {
    // Array de primos (todos os elementos inicializados como true)
    vector<bool> primos(n+1, true);

    // Casos especiais: 0 e 1 não são números primos
    primos[0] = primos[1] = false;

    // Iteração de p no intervalo [2, √n]
    for (int p = 2; p*p <= n; p++) {
        // Se p não é primo
        if (!primos[p]) {
            // Continuar a iteração
            continue;
        }

        // Iteração de i (múltiplos de p) no intervalo [p*p, n]
        for (int i = p*p; i <= n; i += p) {
            // Definir i como não primo
            primos[i] = false;
        }
    }

    // Retornar o array
    return primos;
}

File: src/matematica/numeros_primos/crivo_eratostenes/crivo_eratostenes.py
# -------------------------------------------------------------------
# CRIVO DE ERATÓSTENES
# -------------------------------------------------------------------

# Função para gerar a lista de primos de 0 até n
def crivo(n: int) -> list[bool]:
    # Lista de números (todos os elementos inicializados como True).
    # Cada posição p é um valor booleano que determina se o número
    # p é primo (True) ou não (False).
    primos = [True for i in range(n+1)]

    # Casos especiais: 0 e 1 não são números primos
    primos[0] = primos[1] = False

    # Iteração de p no intervalo [2, √n]
    p = 2
    while p*p <= n:
        # Se p é primo
        if primos[p]:
            # Iteração de i (múltiplos de p) no intervalo [p*p, n]
            for i in range(p*p, n+1, p):
                # Definir i como não primo
                primos[i] = False
        p += 1

    # Retornar a lista
    return primos
